gtk2_ardour/editor_ops.cc

```diff
@@ -3804,65 +3804,79 @@ Editor::trim_region (bool front)
 	commit_reversible_command ();
 }
 
-/** Trim the end of the selected regions to the position of the edit cursor */
+/** Trim the selected regions to the loop range */
 void
 Editor::trim_region_to_loop ()
 {
 	Location* loc = _session->locations()->auto_loop_location();
 	if (!loc) {
 		return;
 	}
-	trim_region_to_location (*loc, _("trim to loop"));
+	trim_region_to_location (*loc, _("trim to loop"), false);
 }
 
+/** Trim the selected regions to the punch range */
 void
 Editor::trim_region_to_punch ()
 {
 	Location* loc = _session->locations()->auto_punch_location();
 	if (!loc) {
 		return;
 	}
-	trim_region_to_location (*loc, _("trim to punch"));
+	trim_region_to_location (*loc, _("trim to punch"), false);
 }
 
 void
-Editor::trim_region_to_location (const Location& loc, const char* str)
+Editor::trim_region_to_location (const Location& loc, const char* str, const bool expand)
 {
 	RegionSelection rs = get_regions_from_selection_and_entered ();
+
+	if (rs.empty()) {
+		return;
+	}
+
 	bool in_command = false;
+	RegionSelection regions_to_remove;
+	/* used when regions are selected but totally out of the range */
 
 	for (RegionSelection::iterator x = rs.begin(); x != rs.end(); ++x) {
 		RegionView* rv = (*x);
 
-		/* require region to span proposed trim */
-		switch (rv->region()->coverage (loc.start(), loc.end())) {
-		case Evoral::OverlapInternal:
-			break;
-		default:
-			continue;
-		}
+		if (!rv->region()->locked()) {
 
-		RouteTimeAxisView* tav = dynamic_cast<RouteTimeAxisView*> (&rv->get_time_axis_view());
-		if (!tav) {
-			return;
-		}
+			rv->region()->clear_changes ();
 
-		samplepos_t start;
-		samplepos_t end;
+			if (rv->region()->first_sample() < loc.start() || (expand && rv->region()->first_sample() > loc.start())) {
```

Comment:
Instead of custom comparisions, it would be preferable to use Region::coverage()  and check for `OverlapNone`

With the previous code's switch statement it might be as easy as
```
case Evoral::OverlapNone:
  rv->region()->playlist()->remove_region (rv->region());
  continue;
```

Reply:
It should have indeed. It relates to bug [7275](https://tracker.ardour.org/view.php?id=7275). In the current master, "Trim to [punch|loop]" does not trim any region if the region does not overlap both the range's start and end, hence a region that is partly in the range is not trimmed at all. Plus, a region that is selected and not in the range at all is left untouched, and I think the user should be trusted to know what he is doing, so if he trims a group of regions and one region is outside the range, this region should be destroyed. That is what this patch does.

Reply:
I've also added a parameter ("expand") which is unused atm, that allows for a "fit to [loop|range]", cutting what exceeds the range and (trying to) expand the regions inside the range to fill the range.